Load real data to the availability page

olena-kozak opened the pull request:
Case if user don't have own slots:
![image](https://user-images.githubusercontent.com/98499085/184256138-51d4942a-a0f9-4134-be65-e70ce00c82bf.png)
Case if user have own slots, but don't have slots connected with teams:
![image](https://user-images.githubusercontent.com/98499085/184256152-8416b19a-21cc-4ce5-a27d-d3220d6ab775.png)
Case when IsVisible = false: 
![image](https://user-images.githubusercontent.com/98499085/184256163-253d4953-045e-4cfb-970f-54e91c720a27.png)
Response when getting teams' slots by user:
![image](https://user-images.githubusercontent.com/98499085/184256390-da8f20b8-8620-4862-9659-b7a4d511a992.png)
Response when getting all available user slots:
![image](https://user-images.githubusercontent.com/98499085/184256403-e985691e-fe29-40e3-ba5e-50eab3a2e1be.png)

olena-kozak committed:
Create IAvailabilityService, AvailabilityService, AvailabilityController

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

# Conflicts:
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Interfaces/IAvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Create availiability-slot, location, team, user models

olena-kozak committed:
Add method to interface IAvailabilityService

olena-kozak committed:
Add base getAvailabilitySlots method

olena-kozak committed:
Add method call to constructor

olena-kozak committed:
Add DTOs and MapperProfiles
Create base request returning entities

olena-kozak committed:
Add SlotTypeProfile
Fix bug with mapping enum

olena-kozak committed:
Change method name

olena-kozak committed:
Change mathod name in AvailabilityService
Add constrtaint to get only slots where slot type equals to "Team"

olena-kozak committed:
Remove comments

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

# Conflicts:
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Interfaces/IAvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/AvailabilityProfile.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs

olena-kozak committed:
Fix bug with mapping AvailabilitySlot

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Change url path to fix secure connection bug
Fix getting slots

olena-kozak committed:
Add logic to display team component with default data
Fix bug with displaying disabled slot

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Fix getting data from backend logic
Create TeamWithSlot dto

olena-kozak committed:
Add methods to group slots by team

olena-kozak committed:
Change displaying data in slot component

olena-kozak committed:
Remove slots-data-helper file

olena-kozak committed:
Create GetAllUserAvailabilitySlotsAsync method to get all users slots

olena-kozak committed:
Add filter to get slots of team, where user is a member

olena-kozak committed:
Add UserSlotComponent for displaying user slots

olena-kozak committed:
Add temporary method to get current user to display data about current user
Add hasOwner check to slot.component to switch between displaying owner name and team member
Add logic to transfer data from availability-page.component.html to user-slot.component.html

olena-kozak committed:
Add temporary method to get current user
Add Id field to UserDto

olena-kozak committed:
Fix bug with filtering user slots by slot type and getting nested objects

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Add visibility icon

olena-kozak committed:
Transfer getCurrentUser method from AvailabilityService to UserService

olena-kozak committed:
Refactor code

olena-kozak committed:
Refactor code

olena-kozak committed:
Fix bug with displaying copy icon during loading user slots

olena-kozak committed:
Remove comments, return http link

olena-kozak committed:
Remove comments, return http link

olena-kozak committed:
Fix bug after EsLint and Prettier check

tatianahutii commented on backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:
```diff
@@ -1,29 +1,59 @@
+        public AvailabilityService(EasyMeetsCoreContext context, IMapper mapper) : base(context, mapper) { }
+        public async Task<ICollection<AvailabilitySlotsGroupByTeamsDto>> GetAllAvailabilitySlotsGroupByTeamsAsync(int id)
         {
+            var teamsWithSlots = await _context.Teams
+                .Include(x => x.AvailabilitySlots)
+                    .ThenInclude(x => x.Members)
+                .Include(x => x.AvailabilitySlots)
+                    .ThenInclude(x => x.Location)
+                .Include(x => x.AvailabilitySlots)
+                    .ThenInclude(x => x.Author)
+                .Where(x => x.AvailabilitySlots.Any(x => x.Type == SlotType.Team))
+                .Where(x => x.AvailabilitySlots.Any(x => x.Members.Any(x => x.UserId == id)))
+                .ToListAsync();
```
This is not a good idea to load all data and only after that map it to a DTO
It would be better to do a projection here and load only necessary fields

tatianahutii commented on backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:
```diff
@@ -1,29 +1,59 @@
 
+        public async Task<ICollection<AvailabilitySlotDto>> GetAllUserAvailabilitySlotsAsync(int id)
+        {
+            var userSlots = await _context.Users
+                 .Where(x => x.Id == id)
+                 .Include(x => x.CreatedSlots)
+                    .ThenInclude(x => x.Members)
+                    .ThenInclude(x => x.User)
+                 .Include(x => x.CreatedSlots)
+                    .ThenInclude(x => x.Location)
+                 .Where(x => x.CreatedSlots.Any(x => x.Type == SlotType.Personal))
+                 .SelectMany(x => x.CreatedSlots)
+                 .ToListAsync();
```
See the previous comment

tatianahutii commented on backend/EasyMeets.Core/EasyMeets.Core.Common/Enums/SlotType.cs:
```diff
@@ -2,6 +2,6 @@ namespace EasyMeets.Core.Common.Enums;
 {
-    Personal,
-    Team
+    Personal = 1,
+    Team = 2
```
Why did you change enum numbering?

olena-kozak replied:
I changed enum numbering because slot type in table is saved as value 1 and 2 (maybe this error was made during seeding). 
Because of that, there was error during loading data from database. To remove this error and convenient enum mapping, I set the value as it is in db. 
So, I made it so that the enum value matched the values in the database and avoided adding (+1).

tatianahutii replied:
It would be better to fix seed and db values

olena-kozak replied:
So, enum value in db should start with 0?

olena-kozak replied:
Can I fix this bug in separate branch, I think it doesn't refer to this task and so that every will know what happened and load new db after fixing seed?

tatianahutii replied:
Yes, sure
Could you please create a card for this?

olena-kozak replied:
Of course. 
So, I commit with wrong (Personal=1) data, so that my page will receive data. And in another brunch I will fix seed and change enum numbering, is it right?

tatianahutii commented on frontend/src/app/core/models/availiability-slot.ts:
```diff
@@ -0,0 +1,13 @@
+    isEnabled: boolean
+    isVisible: boolean
+    location: Location
+    members: Array<User>
```
```suggestion
    members: User[]
```

tatianahutii commented on frontend/src/app/core/models/team-with-slot.ts:
```diff
@@ -0,0 +1,7 @@
+export interface TeamWithSlots {
+    name: string
+    pageLink: string
+    availabilitySlots: Array<AvailabilitySlot>
```
```suggestion
    availabilitySlots: AvailabilitySlot[]
```

tatianahutii commented on frontend/src/app/modules/availability/availability-page/availability-page.component.ts:
```diff
@@ -1,8 +1,57 @@
+
+    public currentUser: User;
+
+    public userSlots: Array<AvailabilitySlot>;
```
```suggestion
    public userSlots: AvailabilitySlot[];
```

tatianahutii commented on frontend/src/app/modules/availability/availability-page/availability-page.component.ts:
```diff
@@ -1,8 +1,57 @@
+
+    public getCurrentUser() {
+        this.userService
+            .getCurrentUserById(4)
```
It would be better to not hard code the number
At least we need to move it to a separate variable

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'development' into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'feature/load_data_to_availability_page' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Remove hard coded current user id, add separate variable for it
Change way of type assignment (array type)

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

# Conflicts:
#	frontend/src/app/modules/availability/availability.module.ts

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

# Conflicts:
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Interfaces/IUserService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/UserProfile.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/UserService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.Common/DTO/User/UserDto.cs
#	backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/UserController.cs

olena-kozak committed:
Change link path

olena-kozak committed:
Fix bug with displaying avatars in slots and userName i user-slot.component.html

olena-kozak committed:
Remove unnecessary enum value

olena-kozak committed:
Remove automapping after loading all data:
add projection in GetAllAvailabilitySlotsGroupByTeamsAsync method (AvailabilityService) to load only necessary fields

tatianahutii commented on backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:
```diff
@@ -21,13 +23,52 @@ public async Task<ICollection<AvailabilitySlotsGroupByTeamsDto>> GetAllAvailabil
+                    {
+                        Name = x.Name,
+                        PageLink = x.PageLink,
+                        AvailabilitySlots = x.AvailabilitySlots
+                            .Select(y =>
+                                new AvailabilitySlotDto
+                                {
+                                    Id = y.Id,
+                                    Name = y.Name,
+                                    Link = y.Link,
+                                    Type = y.Type,
+                                    Size = y.Size,
+                                    IsEnabled = y.IsEnabled,
+                                    Author =
+                                        new UserDto
+                                        {
+                                            Id = y.Author.Id,
+                                            UserName = y.Author.Name,
+                                            Email = y.Author.Email,
+                                            Image = y.Author.ImagePath,
+                                            Phone = y.Author.PhoneNumber
+                                        },
+                                    Location = new LocationDto
+                                    {
+                                        Name = y.Location.Name,
+                                    },
+                                    Members = y.Members
+                                        .Select(m =>
+                                         new UserDto
+                                         {
+                                             Id = m.User.Id,
+                                             UserName = m.User.Name,
+                                             Email = m.User.Email,
+                                             Image = m.User.ImagePath,
+                                             Phone = m.User.PhoneNumber
+                                         })
+                                        .ToList()
+                                })
+                            .ToList()
+                    })
```
Are you really need all this field?
You only need to load data that you will use on your page

olena-kozak replied:
Yes, I need these data, I can remove user email and phone number of user (maybe also link in availability slot for now, but it will be used later, so I decided to leave it).
But other fields are displaying on availability page.
![image](https://user-images.githubusercontent.com/98499085/184328118-32d53f5c-52b2-4617-8dcb-1fd93407187e.png)

olena-kozak committed:
Remove unnecessary fields used during loading data from db

olena-kozak committed:
Remove automapping after loading all data:
add projection in GetAllUserAvailabilitySlotsAsync method (AvailabilityService) to load only necessary fields

olena-kozak committed:
Remove unused field in User interface

olena-kozak committed:
Add migration to update slot type values
Set new rule for faker to generate type for availability slot in GenerateAvailabilitySlots method in ModelBuilderExtensions

olena-kozak committed:
Change link
Change database snapshot after updating database

olena-kozak committed:
Change url link

tatianahutii reviewed (changes requested):
Could you please fix warnings?
![image](https://user-images.githubusercontent.com/38536017/184365316-a14c8aaa-0bfe-4d3d-8d31-b7c915559d96.png)

olena-kozak committed:
Merge branch 'development' into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'development' into bug/wrong_slot_type_value_in_db

olena-kozak committed:
Fix warning with field Members in AvailabilitySlotsGroupByTeamsDto

olena-kozak committed:
Merge branch 'feature/load_data_to_availability_page' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Fix bug with enum values in SlotType

tatianahutii reviewed (changes requested):
Still see a few git warnings

olena-kozak committed:
Fix warning with field Members in AvailabilitySlotsGroupByTeamsDto

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

# Conflicts:
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/EasyMeets.Core.BLL.csproj
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Interfaces/IAvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/AvailabilityProfile.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs

tatianahutii commented on backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:
```diff
@@ -1,23 +1,119 @@
+                 .Where(x => x.CreatedSlots.Any(x => x.Type == SlotType.Personal))
+                 .SelectMany(x => x.CreatedSlots)
+                 .Select(y =>
+                    new AvailabilitySlotDto
+                    {
+                        Id = y.Id,
+                        Name = y.Name,
+                        Type = y.Type,
+                        Size = y.Size,
+                        IsEnabled = y.IsEnabled,
+                        Author =
+                            new UserDto
+                            {
+                                UserName = y.Author.Name,
+                                Image = y.Author.ImagePath,
+                            },
+                        Location = 
+                            new LocationDto
+                            {
+                                Name = y.Location.Name,
+                            },
+                        Members = y.Members
+                            .Select(m =>
+                                new UserDto
+                                {
+                                    UserName = m.User.Name,
+                                    Image = m.User.ImagePath,
+                                })
+                            .ToList()
+                    })
```
Could you simplify this to one layer DTO?

olena-kozak replied:
Yes.
And leave projection here or use mapper?

olena-kozak replied:
I simplified dto and use projection in this method, so that unused data will not be load in this method

tatianahutii commented on backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs:
```diff
@@ -1,23 +1,119 @@
+                                    Size = y.Size,
+                                    IsEnabled = y.IsEnabled,
+                                    Author =
+                                        new UserDto
+                                        {
+                                            UserName = y.Author.Name,
+                                            Image = y.Author.ImagePath,
+                                        },
+                                    Location = new LocationDto
+                                    {
+                                        Name = y.Location.Name,
+                                    },
+                                    Members = y.Members
+                                        .Select(m =>
+                                         new UserDto
+                                         {
+                                             UserName = m.User.Name,
+                                             Image = m.User.ImagePath,
+                                         })
+                                        .ToList()
```
You can create simple DTO and use mapper here

olena-kozak replied:
I simplified dto and used mapper in this method

tatianahutii reviewed (changes requested):
Please use icon instead of png
![image](https://user-images.githubusercontent.com/38536017/184476789-4371ad55-af23-4091-8268-68c7792adaa9.png)

olena-kozak committed:
Change visibility image to icon in slot component

olena-kozak committed:
Remove visibility image from assets

olena-kozak committed:
Simplify AvailabilitySlotDto to one layer DTO
Fix dto in GetAllAvailabilitySlotsGroupByTeamsAsync method (projection is left)
Add AvailabilitySlotMemberDto

olena-kozak committed:
Use simplified AvailabilitySlotDto in GetAllAvailabilitySlotsGroupByTeamsAsync method
Use mapper

olena-kozak committed:
Add availability-slot-member dto
Simplify availiability-slot interface
Fix bug in slot.component.html

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'development' into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into bug/wrong_slot_type_value_in_db

# Conflicts:
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/EasyMeets.Core.BLL.csproj
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Interfaces/IAvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/AvailabilityProfile.cs
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs
#	backend/EasyMeets.Core/EasyMeets.Core.DAL/Migrations/EasyMeetsCoreContextModelSnapshot.cs
#	backend/EasyMeets.Core/EasyMeets.Core.WebAPI/Controllers/AvailabilityController.cs

olena-kozak committed:
Merge branch 'feature/load_data_to_availability_page' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into bug/wrong_slot_type_value_in_db

# Conflicts:
#	backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/AvailabilityProfile.cs

olena-kozak committed:
Fix bug with filtering AvailabilitySlotsGroupByTeams

olena-kozak committed:
Merge branch 'bug/wrong_slot_type_value_in_db' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Change wrong link
Fix bug with filtering AllUserAvailabilitySlots

olena-kozak committed:
Fix displaying visibility icon

olena-kozak committed:
Merge branch 'development' into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Fix bug with making two separate requests on backend
Add method GetUserPersonalAndTeamSlotsAsync to get all user personal and team slots
Create UserPersonalAndTeamSlots dto

olena-kozak committed:
Merge branch 'feature/load_data_to_availability_page' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Rename UserPersonalAndTeamSlotsDto

olena-kozak committed:
Remove methods making making two separate requests on frontend
Add method getAllUsersAvailabilitySlots to get all user personal and team slots
Create user-personal-and-team-slots.ts, availability-slots-group-by-team.ts

olena-kozak committed:
Rename interfaces
Remove unnecessary variables

olena-kozak committed:
Fix bug with mapping dto in dto in GetUserPersonalAndTeamSlotsAsync method

olena-kozak committed:
Fix bugs ci check with import after

tatianahutii reviewed (approved):
LGTM

tatianahutii reviewed (changes requested):
Need recreate megration

olena-kozak committed:
Merge branch 'development' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

olena-kozak committed:
Merge branch 'development' into feature/load_data_to_availability_page

olena-kozak committed:
Remove Update_SlotType_Values_In_AvailabilitySlot_Table migration
Revert SlotType value

olena-kozak committed:
Merge branch 'feature/load_data_to_availability_page' of https://github.com/BinaryStudioAcademy/bsa-2022-easymeets into feature/load_data_to_availability_page

tatianahutii reviewed (approved):
LGTM
